javascript/ql/src/semmle/javascript/MapAndSet.qll

```diff
@@ -0,0 +1,279 @@
+   */
+  private class SetConstructor extends MapOrSetFlowStep, DataFlow::NewNode {
+    SetConstructor() { this = DataFlow::globalVarRef("Set").getAnInstantiation() }
+
+    override predicate loadStore(
+      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
+    ) {
+      pred = this.getArgument(0) and
+      succ = this and
+      fromProp = arrayLikeElement() and
+      toProp = setElement()
+    }
+  }
+
+  /**
+   * A step for a `for of` statement on a Map, Set, or Iterator.
+   * For Sets and iterators the l-value are the elements of the set/iterator.
+   * For Maps the l-value is a tuple containing a key and a value.
+   *
+   * This is partially duplicated behavior with the `for of` step for Arrays (in Arrays.qll).
+   * This duplication is required for the type-tracking steps defined in `MapsAndSetsTypeTracking`.
+   */
+  private class ForOfStep extends MapOrSetFlowStep, DataFlow::ValueNode {
+    ForOfStmt forOf;
+    DataFlow::Node element;
+
+    ForOfStep() {
+      this.asExpr() = forOf.getIterationDomain() and
+      element = DataFlow::lvalueNode(forOf.getLValue())
+    }
+
+    override predicate load(DataFlow::Node obj, DataFlow::Node e, string prop) {
+      obj = this and
+      e = element and
+      prop = arrayLikeElement()
+    }
+
+    override predicate loadStore(
+      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
+    ) {
+      pred = this and
+      succ = element and
+      fromProp = mapValueUnknownKey() and
+      toProp = "1"
+    }
+  }
+
+  /**
+   * A step for a call to `forEach` on a Set or Map.
+   */
+  private class SetMapForEach extends MapOrSetFlowStep, DataFlow::MethodCallNode {
+    SetMapForEach() { this.getMethodName() = "forEach" }
+
+    override predicate load(DataFlow::Node obj, DataFlow::Node element, string prop) {
+      obj = this.getReceiver() and
+      element = this.getCallback(0).getParameter(0) and
+      prop = [setElement(), mapValueUnknownKey()]
+    }
+  }
+
+  /**
+   * A call to the `get` method on a Map.
+   * If the key of the call to `get` has a known string value, then only the value corresponding to that key will be retrieved.
+   */
+  private class MapGet extends MapOrSetFlowStep, DataFlow::MethodCallNode {
+    MapGet() { this.getMethodName() = "get" }
+
+    override predicate load(DataFlow::Node obj, DataFlow::Node element, string prop) {
+      obj = this.getReceiver() and
+      element = this and
+      prop = mapValue(this.getArgument(0))
+    }
+
+    override predicate canLoadKnownKey() { any() }
+  }
+
+  /**
+   * A call to the `set` method on a Map.
+   *
+   * If the key of the call to `set` has a known string value,
+   * then the value will be saved into a pseudo-property corresponding to the known string value.
+   * The value will additionally be saved into a pseudo-property corresponding to values with unknown keys.
```

Comment:
Performance consideration: what if we instead do the join of known and unknown values on reads instead (like in TAJS)?

Reply:
The current name of `mapValuesUnknownKey` is misleading, since I always write to that pseudo-property, even when the key is known. 

A more sound approach like TAJS sounds like a good idea.
I think I'll do that by splitting fuzzy read/writes on maps into 2 different pseudo-properties. 

- `allMapValues`, which is always written to, and used in `ForOfStep` and similar where we need all values independently of whether the key is known/unknown. This is basically the current behavior of the `mapValueUnknownKey` pseudo-property. 
- `mapValuesUnknownKey`, which is only only used when the key is unknown. 

By using the `allMapValues` I can avoid a loadStep that enumerates all known properties. 

It will also make the disjunct you pointed out simpler.